include/rspq.h

```diff
@@ -0,0 +1,704 @@
+ * to respectively do a single check or block waiting for the syncpoint to be
+ * reached by RSP.
+ * 
+ * Syncpoints are implemented using interrupts, so they have a light but non
+ * trivial overhead. Do not abuse them. For instance, it is reasonable to use
+ * tens of syncpoints per frame, but not hundreds or thousands of them.
+ * 
+ * @note A valid syncpoint is an integer greater than 0.
+ */
+typedef int rspq_syncpoint_t;
+
+/**
+ * @brief Initialize the RSPQ library.
+ * 
+ * This should be called by the initialization functions of the higher-level
+ * libraries using the RSP command queue. It can be safely called multiple
+ * times without side effects.
+ * 
+ * It is not required by applications to call this explicitly in the main
+ * function.
+ */
+void rspq_init(void);
+
+/**
+ * @brief Shut down the RSPQ library.
+ * 
+ * This is mainly used for testing.
+ */
+void rspq_close(void);
+
+
+/**
+ * @brief Register a rspq overlay into the RSP queue engine.
+ * 
+ * This function registers a rspq overlay into the queue engine.
+ * An overlay is a RSP ucode that has been written to be compatible with the
+ * queue engine (see rsp_queue.inc) and is thus able to execute commands
+ * that are enqueued in the queue. An overlay doesn't have a single entry point:
+ * it exposes multiple functions bound to different commands, that will be
+ * called by the queue engine when the commands are enqueued.
+ * 
+ * The function returns the overlay ID, which is the ID to use to enqueue
+ * commands for this overlay. The overlay ID must be passed to #rspq_write
+ * when adding new commands. rspq allows up to 16 overlays to be registered
+ * simultaneously, as the overlay ID occupies the top 4 bits of each command.
+ * 
+ * Notice that it is possible to occupy multiple consecutive IDs with the
+ * same overlay in case the overlay exposes more than 16 commands. For instance,
+ * registering an overlay that handles up to 32 commands could return ID 0x6, 
+ * meaning that the whole range of command IDs 0x60-0x7F would be assigned to it.
+ *             
+ * @param      overlay_ucode  The overlay to register
+ *
+ * @return     The overlay ID that has been assigned to the overlay. Note that
+ *             this value will be preshifted by 28 (eg: 0x60000000 for ID 6),
+ *             as this is the expected format used by #rspq_write.
+ */
+uint32_t rspq_overlay_register(rsp_ucode_t *overlay_ucode);
+
+/**
+ * @brief Register an overlay into the RSP queue engine assigning a static ID to it
+ * 
+ * This function works similar to #rspq_overlay_register, except it will
+ * attempt to assign the specified ID to the overlay instead of automatically
+ * choosing one. Note that if the ID (or a consecutive IDs) is already used
+ * by another overlay, this function will assert, so careful usage is advised.
+ * 
+ * Assigning a static ID can mostly be useful for debugging purposes.
```

Comment:
yes that's the idea. You initialized your own libraries first to reserve the fixed IDs. Or, knowing that rspq allocates in growing order, you can simply reserve as static the last IDs (0xF, 0xE, etc.) so that will work independently from libdragon's internal libraries.

Reply:
Also note that for the purpose of forwarding RDP commands, IDs 0x2 and 0x3 are "soft reserved", which means when registering overlays, IDs 0x2 and 0x3 will be assigned last until no others are available. The upcoming "gfx" overlay will statically register itself with those IDs. But if gfx is not actually needed, those IDs are still available for other custom overlays.